File: ui/paper_code_audit_viewer.js
"use strict";
const app = document.getElementById("app");
const esc = (v) =>
  String(v ?? "").replace(
    /[&<>"']/g,
    (c) =>
      ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" })[
        c
      ],
  );
const fmt = (n) => Number(n).toLocaleString();
const labels = {
  needed: "Needed",
  "protected-other-git-worktree": "Other Git worktree",
  "conditional-retirement-candidate": "Conditional candidate",
  "unresolved-do-not-delete": "Unresolved · keep",
};
const badge = (s) =>
  `<span class="badge ${esc(s)}">${esc(labels[s] || s)}</span>`;
const link = (id) =>
  `<a class="badge" href="#audit/${encodeURIComponent(id)}">${esc(id)}</a>`;
const links = (ids) =>
  `<div class="auditlinks">${String(ids).split(";").filter(Boolean).map(link).join("")}</div>`;
const download = (name, label) =>
  `<a class="button" href="/download?name=${encodeURIComponent(name)}">${esc(label)}</a>`;
const heading = (tag, title, desc) =>
  `<div class="eyebrow">${esc(tag)}</div><h1>${esc(title)}</h1><p class="muted">${esc(desc)}</p>`;
const table = (headers, rows) =>
  `<div class="tablewrap"><table><thead><tr>${headers.map((h) => `<th>${esc(h)}</th>`).join("")}</tr></thead><tbody>${rows.length ? rows.map((r) => `<tr>${r.map((c) => `<td>${c}</td>`).join("")}</tr>`).join("") : `<tr><td colspan="${headers.length}">No matching records.</td></tr>`}</tbody></table></div>`;
const renderText = (text) =>
  esc(text)
    .split("\n")
    .map((l) =>
      l.startsWith("### ")
        ? `<h3>${l.slice(4)}</h3>`
        : l.startsWith("## ")
          ? `<h2>${l.slice(3)}</h2>`
          : l.startsWith("# ")
            ? `<h1>${l.slice(2)}</h1>`
            : /^\s*- /.test(l)
              ? `<div class="bullet">${l.replace(/^(\s*)- /, "$1• ")}</div>`
              : l,
    )
    .join("\n")
    .replace(/\*\*([^*\n]+)\*\*/g, "<strong>$1</strong>")
    .replace(/`([^`\n]+)`/g, "<code>$1</code>");
async function api(path, params = {}) {
  const r = await fetch("/api/" + path + "?" + new URLSearchParams(params));
  if (!r.ok) throw Error(`HTTP ${r.status}: ${await r.text()}`);
  return r.json();
}
let epoch = 0;
function filters() {
  return `<input id="search" aria-label="Search paths" placeholder="Search full path, filename or folder…"><select id="status" aria-label="File status"><option value="">All statuses</option>${Object.entries(
    labels,
  )
    .map(([k, v]) => `<option value="${k}">${v}</option>`)
    .join("")}</select>`;
}

async function overview(token) {
  const s = await api("status");
  if (token !== epoch) return;
  const c = s.status_counts;
  app.innerHTML =
    heading(
      "Audit overview",
      "What can we safely keep or remove?",
      "45 agent audits across three waves. Explore each result, its dependencies and the open questions.",
    ) +
    `<div class="notice"><strong>This is not a deletion certificate.</strong> ${fmt(c["unresolved-do-not-delete"])} inventoried code files remain unresolved. Source candidates still require a retirement decision.</div>${s.changed_inputs_since_load.length ? `<div class="notice error">Audit inputs changed after server startup: ${esc(s.changed_inputs_since_load.join(", "))}. Restart the viewer to review the new records.</div>` : ""}<div class="cards"><div class="card"><b>${fmt(c.needed)}</b><span>Code files with retention evidence</span></div><div class="card"><b>${fmt(c["unresolved-do-not-delete"])}</b><span>Unresolved code files</span></div><div class="card"><b>${s.source_candidates}</b><span>Conditional source candidates</span></div><div class="card"><b>${fmt(s.needed_files)}</b><span>Concrete files on the keep list</span></div></div><div class="grid"><section class="panel"><h2>Code inventory</h2><p class="muted">${fmt(s.code_files)} code/configuration filenames in the original snapshot.</p><div class="bar">${Object.entries(
      c,
    )
      .map(
        ([k, n]) =>
          `<div class="${esc(k)}" style="width:${(100 * n) / s.code_files}%" title="${esc(labels[k])}: ${n}"></div>`,
      )
      .join("")}</div><div class="legend">${Object.entries(c)
      .map(([k, n]) => `<div>${badge(k)}<strong>${fmt(n)}</strong></div>`)
      .join(
        "",
      )}</div></section><section class="panel"><h2>Coverage and limits</h2><p><strong>${s.figures} figures · ${s.tables} tables · ${s.agent_audits} audits</strong></p><p>${s.directory_paths} directory paths also have explicit subset-selection rules. Their entire contents are not certified as needed.</p><p>Runtime, result reproduction and historical provenance are separate reasons to retain a file.</p><p><a href="#review">Review the new consistency checks →</a></p><p><a href="#audits">Browse all experiment assignments →</a></p><p class="reviewstamp">Loaded ${esc(s.loaded_at)}<br>Read-only server · consistency checks ${s.review_passed ? "passed" : "need attention"}</p></section></div><section class="panel"><h2>Files that look old can still be required</h2><p>Current results use dated review scripts, recovery archives and old analysis modules. The audit recovered 27 launch/configuration files inside an earlier deletion archive.</p>${link("root_recovered_archives")} ${link("root_import_closure")}</section><div class="toolbar">${download("report.md", "Download full report")}${download("needed_files.csv", "Download keep list")}${download("code_status.csv", "Download code status")}</div>`;
}
async function audits(token) {
  const rows = await api("audits");
  if (token !== epoch) return;
  app.innerHTML =
    heading(
      "Result provenance",
      "Experiment audits",
      "One report per assigned result family or supporting method. Open an audit to see its findings and exact file evidence.",
    ) +
    `<div class="toolbar"><input id="auditsearch" aria-label="Search audits" placeholder="Search game, model, method or audit ID…"></div><div id="results"></div>`;
  const draw = () => {
    const q = document.getElementById("auditsearch").value.toLowerCase();
    document.getElementById("results").innerHTML = table(
      ["Audit", "Result / method", "Needed entries", "Open questions"],
      rows
        .filter((r) => JSON.stringify(r).toLowerCase().includes(q))
        .map((r) => [
          link(r.id),
          esc(r.result),
          fmt(r.needed_entries),
          fmt(r.open_questions),
        ]),
    );
  };
  document.getElementById("auditsearch").oninput = draw;
  draw();
}
async function audit(id, token) {
  const d = await api("audit", { id });
  if (token !== epoch) return;
  app.innerHTML =
    heading(
      id,
      d.result,
      "Positive dependency evidence and unresolved questions from this assignment.",
    ) +
    `<section class="panel"><h2>Finding</h2><p>${esc(d.summary)}</p></section><section class="panel"><h2>Open questions</h2>${d.unresolved.length ? `<ul>${d.unresolved.map((x) => `<li>${esc(typeof x === "string" ? x : JSON.stringify(x))}</li>`).join("")}</ul>` : "<p>No questions recorded by this audit; this is not a global completeness claim.</p>"}</section><div class="toolbar">${download(d.report_name, "Download this report")}<a class="button" href="#files?audit=${encodeURIComponent(id)}">Explore listed files</a></div><details><summary>Dependency entries (${d.needed.length})</summary>${table(
      ["Path", "Role", "Reason / evidence"],
      d.needed
        .slice(0, 150)
        .map((r) => [
          `<span class="path">${esc(r.path)}</span>`,
          esc(r.role),
          esc(r.reason) + "<br><small>" + esc(r.evidence) + "</small>",
        ]),
    )}${d.needed.length > 150 ? "<p>Showing 150 entries here. Use the file explorer or full dependency download for all entries.</p>" : ""}</details><section class="panel report">${renderText(d.report)}</section>`;
}
async function files(query, token) {
  const params = new URLSearchParams(query),
    audit = params.get("audit") || "";
  app.innerHTML =
    heading(
      "File-level evidence",
      "File explorer",
      "Search full paths. Click a file to see the reports and reasons that support its status.",
    ) +
    `<div class="toolbar"><select id="dataset" aria-label="Inventory"><option value="code">Code inventory (1,204)</option><option value="needed" ${audit ? "selected" : ""}>Concrete keep list (12,078)</option></select>${filters()}</div>${audit ? `<p>Filtered to ${link(audit)} · <a href="#files">Clear audit filter</a></p>` : ""}<div id="results"></div><div class="pager"><button id="prev">Previous</button><span id="range"></span><button id="next">Next</button></div>`;
  let offset = 0,
    total = 0,
    request = 0;
  const draw = async () => {
    const mine = ++request;
    const dataset = document.getElementById("dataset").value;
    document.getElementById("status").disabled = dataset === "needed";
    const data = await api("files", {
      dataset,
      q: document.getElementById("search").value,
      status: dataset === "code" ? document.getElementById("status").value : "",
      audit,
      offset,
      limit: 100,
    });
    if (token !== epoch || mine !== request) return;
    total = data.total;
    document.getElementById("results").innerHTML = table(
      ["Full path", dataset === "code" ? "Status" : "Retention role", "Audits"],
      data.rows.map((r) => [
        `<span class="path"><a href="#file/${encodeURIComponent(r.path)}">${esc(r.path)}</a></span>`,
        dataset === "code" ? badge(r.status) : esc(r.roles),
        links(r.audit_ids),
      ]),
    );
    document.getElementById("range").textContent =
      `${total ? offset + 1 : 0}–${Math.min(offset + 100, total)} of ${fmt(total)}`;
    document.getElementById("prev").disabled = offset === 0;
    document.getElementById("next").disabled = offset + 100 >= total;
  };
  const change = () => {
    offset = 0;
    draw().catch(showError);
  };
  document.getElementById("search").oninput = change;
  document.getElementById("dataset").onchange = change;
  document.getElementById("status").onchange = change;
  document.getElementById("prev").onclick = () => {
    offset = Math.max(0, offset - 100);
    draw().catch(showError);
  };
  document.getElementById("next").onclick = () => {
    offset += 100;
    draw().catch(showError);
  };
  await draw();
}
async function file(path, token) {
  const d = await api("file", { path });
  if (token !== epoch) return;
  app.innerHTML =
    heading(
      "Decision evidence",
      "File detail",
      "This viewer shows audit evidence, not file contents.",
    ) +
    `<div class="codebox path">${esc(path)}</div><p>${d.code ? badge(d.code.status) : badge(d.keep ? "needed" : "conditional-retirement-candidate")} · Path ${d.exists_now ? "exists now" : "is missing now"}</p>${d.candidate ? `<section class="panel"><h2>Candidate rationale</h2><p>${esc(d.candidate.reason)}</p><p><strong>No deletion is authorized.</strong></p>${link(d.candidate.audit)}</section>` : ""}${
      !d.evidence.length
        ? '<div class="notice">No positive file-level dependency evidence is recorded here. This does not prove the file is unused.</div>'
        : table(
            ["Audit", "Role", "Reason", "Source evidence"],
            d.evidence.map((r) => [
              link(r.audit),
              esc(r.role),
              esc(r.reason),
              `<span class="path">${esc(r.evidence)}</span>`,
            ]),
          )
    }`;
}
async function candidates(token) {
  const rows = await api("candidates");
  if (token !== epoch) return;
  app.innerHTML =
    heading(
      "Conditional retirement",
      "Cleanup candidates",
      "Seven generic planning helpers, two older plotting frontends, and 25 generated test-bytecode files.",
    ) +
    `<div class="notice"><strong>There are no delete or approval controls.</strong> Source candidates require confirmation that their standalone workflows are no longer needed. Old plot scripts are a separate decision from runtime cleanup.</div><div class="toolbar"><select id="kind" aria-label="Candidate type"><option value="candidate-only">Source candidates (9)</option><option value="safe-generated">Generated bytecode (25)</option><option value="">All candidates</option></select>${download("cleanup_candidates.json", "Download evidence")}</div><div id="results"></div>`;
  const draw = () => {
    const kind = document.getElementById("kind").value;
    document.getElementById("results").innerHTML = table(
      ["Exact path", "Reason", "Audit"],
      rows
        .filter((r) => !kind || r.confidence === kind)
        .map((r) => [
          `<span class="path"><a href="#file/${encodeURIComponent(r.path)}">${esc(r.path)}</a></span>`,
          esc(r.reason),
          link(r.audit),
        ]),
    );
  };
  document.getElementById("kind").onchange = draw;
  draw();
}
async function coverage(token) {
  const rows = await api("coverage");
  if (token !== epoch) return;
  app.innerHTML =
    heading(
      "Paper → result → files",
      "Paper coverage",
      "All 30 figures and 10 numbered tables, linked to their audit assignments and source locations.",
    ) +
    table(
      ["Paper item", "Label / source", "Audits", "Graphic inputs"],
      rows.map((r) => [
        `${esc(r.kind)} ${esc(r.source_order_number)}`,
        `<strong>${esc(r.labels)}</strong><br><span class="path">${esc(r.source_path)}:${esc(r.line)}</span>`,
        links(r.audit_ids),
        `<span class="path">${esc(r.graphic_inputs || "Table embedded in LaTeX")}</span>`,
      ]),
    ) +
    `<div class="toolbar">${download("paper_coverage.csv", "Download coverage map")}</div>`;
}
async function review(token) {
  const [d, selections] = await Promise.all([api("review"), api("selections")]);
  if (token !== epoch) return;
  app.innerHTML =
    heading(
      "Independent record checks",
      "Review of the audit",
      "Rechecked from the detailed reports and actual filesystem paths when this server started.",
    ) +
    `<div class="notice"><strong>${d.checks.filter((c) => c.passed).length}/${d.checks.length} record checks passed.</strong> Passing these checks does not certify scientific validity, complete dependency coverage, or safe deletion.</div><p class="reviewstamp">Checked ${esc(d.checked_at)}</p><div class="checks">${d.checks.map((c) => `<div class="check ${c.passed ? "" : "fail"}"><strong>${c.passed ? "PASS" : "NEEDS REVIEW"} · ${esc(c.name)}</strong><small>${esc(c.scope)}</small>${c.issues.length ? `<pre class="path">${esc(JSON.stringify(c.issues, null, 2))}</pre>` : ""}</div>`).join("")}</div><section class="panel"><h2>What remains unproven</h2><ul>${d.limits.map((x) => `<li>${esc(x)}</li>`).join("")}</ul></section><details><summary>Directory-selection evidence (${selections.length} records)</summary>${table(
      ["Directory", "Audit", "Selection rule"],
      selections.map((s) => [
        `<span class="path">${esc(s.path)}</span>`,
        link(s.audit),
        esc(s.reason),
      ]),
    )}</details>`;
}
async function report(token) {
  const d = await api("report");
  if (token !== epoch) return;
  app.innerHTML =
    heading(
      "Final audit deliverable",
      "Full report",
      "The original report is preserved. Fresh record checks are available on the Review checks page.",
    ) +
    `<div class="toolbar">${download("report.md", "Download Markdown report")}${download("audit_index.csv", "Download audit index")}</div><section class="panel report">${renderText(d.text)}</section>`;
}
function showError(error) {
  app.innerHTML = `<div class="notice error"><h2>Could not load audit records</h2><p>${esc(error.message)}</p><p>No data was substituted. Check the server and reload this page.</p></div>`;
}
async function route() {
  const token = ++epoch;
  const hash = location.hash.slice(1) || "overview";
  const [path, query = ""] = hash.split("?");
  document
    .querySelectorAll("nav a")
    .forEach((a) =>
      a.classList.toggle(
        "active",
        a.hash ===
          "#" +
            (path.startsWith("audit/")
              ? "audits"
              : path.startsWith("file/")
                ? "files"
                : path),
      ),
    );
  app.innerHTML = '<div class="loading">Loading evidence…</div>';
  try {
    if (path === "overview") await overview(token);
    else if (path === "audits") await audits(token);
    else if (path.startsWith("audit/"))
      await audit(decodeURIComponent(path.slice(6)), token);
    else if (path === "files") await files(query, token);
    else if (path.startsWith("file/"))
      await file(decodeURIComponent(path.slice(5)), token);
    else if (path === "candidates") await candidates(token);
    else if (path === "coverage") await coverage(token);
    else if (path === "review") await review(token);
    else if (path === "report") await report(token);
    else throw Error("Unknown view");
  } catch (e) {
    if (token === epoch) showError(e);
  }
}
window.addEventListener("hashchange", route);
route();

File: ui/paper_code_audit_viewer.py
#!/usr/bin/env python3
"""Read-only audit browser. Serves only indexed audit reports, never source files."""
from __future__ import annotations

import argparse
import csv
import hashlib
import json
import re
import importlib.util
import struct
from collections import Counter, defaultdict
from datetime import datetime, timezone
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlsplit

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_AUDIT = ROOT / "docs/analysis/paper_code_audit_20260907"
ASSETS = Path(__file__).resolve().parent
STATUSES = {"needed", "protected-other-git-worktree", "conditional-retirement-candidate", "unresolved-do-not-delete"}


def short_id(value: str) -> str:
    match = re.match(r"^(e\d{2})(?:_|$)", value)
    return match[1] if match else value


def read_csv(path: Path) -> list[dict]:
    with path.open(newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


class AuditCatalog:
    def __init__(self, directory: Path = DEFAULT_AUDIT):
        self.directory = directory.resolve()
        self.index = read_csv(self.directory / "audit_index.csv")
        self.code = read_csv(self.directory / "code_status.csv")
        self.needed = read_csv(self.directory / "needed_files.csv")
        self.coverage = read_csv(self.directory / "paper_coverage.csv")
        self.candidates = json.loads((self.directory / "cleanup_candidates.json").read_text())
        self.selections = json.loads((self.directory / "directory_selections.json").read_text())
        self.previous = json.loads((self.directory / "validation.json").read_text())
        self.inventory = json.loads((self.directory / "e43_code_filename_inventory.json").read_text())
        self.audits = {}
        self.evidence = defaultdict(list)
        self.documents = {}
        for row in self.index:
            report = Path(row["report_path"]).resolve()
            dependency = Path(row["dependency_report_path"]).resolve()
            for path in (report, dependency):
                if path.parent != self.directory:
                    raise ValueError(f"Report path escapes the audit directory: {path}")
            data = json.loads(dependency.read_text())
            key = short_id(data["task_id"])
            if key in self.audits:
                raise ValueError(f"Duplicate audit ID: {key}")
            self.audits[key] = {**data, "id": key, "report": report.read_text(), "report_name": report.name}
            self.documents[report.name] = report.read_bytes()
            self.documents[dependency.name] = dependency.read_bytes()
            for entry in data["needed"]:
                self.evidence[entry["path"]].append({"audit": key, **entry})
        for name in ("report.md", "needed_files.csv", "code_status.csv", "audit_index.csv", "paper_coverage.csv", "cleanup_candidates.json", "directory_selections.json", "unresolved_questions.json", "validation.json", "e43_code_filename_inventory.json"):
            self.documents[name] = (self.directory / name).read_bytes()
        self.code_by_path = {row["path"]: row for row in self.code}
        self.needed_by_path = {row["path"]: row for row in self.needed}
        self.candidate_by_path = {row["path"]: row for row in self.candidates}
        self.review = self.validate()
        self.loaded_at = datetime.now(timezone.utc).isoformat()
        self.fingerprints = {}
        for name in sorted(self.documents):
            path = self.directory / name
            stat = path.stat()
            self.fingerprints[name] = (stat.st_size, stat.st_mtime_ns)
        self.digest = hashlib.sha256(b"".join(name.encode() + self.documents[name] for name in sorted(self.documents))).hexdigest()

    def validate(self) -> dict:
        checks = []

        def check(name: str, issues: list, scope: str):
            checks.append({"name": name, "passed": not issues, "issues": issues, "scope": scope})

        agents = {key for key in self.audits if re.fullmatch(r"e\d{2}", key)}
        check("All 45 audit assignments have reports", sorted(agents ^ {f"e{i:02}" for i in range(1, 46)}), "Assignment coverage, not proof that every source line was reviewed.")
        check("Keep paths are unique", [p for p, n in Counter(r["path"] for r in self.needed).items() if n != 1], "Exact path strings in the consolidated CSV.")
        concrete = {p for p in self.evidence if Path(p).is_file()}
        check("Keep list matches the detailed reports", sorted(concrete ^ set(self.needed_by_path)), "Union of concrete files; directory selections remain separate.")
        check("All reported dependency paths exist", [p for p in self.evidence if not Path(p).exists()], "File existence does not prove historical byte equality or scientific validity.")
        check("Every code path has one valid status", [r for r in self.code if r["status"] not in STATUSES] + [p for p, n in Counter(r["path"] for r in self.code).items() if n != 1], "No absent status is treated as unused.")
        check("Code inventory matches status rows", sorted(set(self.inventory["paths"]) ^ set(self.code_by_path)), "The original 1,204-file inventory is a dated, extension-limited snapshot.")
        check("Needed code rows have positive evidence", [r["path"] for r in self.code if r["status"] == "needed" and r["path"] not in concrete], "Evidence can describe runtime, reproduction, provenance or support.")
        conflicts = []
        resolved_keep = {str(Path(p).resolve()) for p in self.needed_by_path}
        resolved_directories = {Path(s["path"]).resolve() for s in self.selections}
        for candidate in self.candidates:
            path = Path(candidate["path"]).resolve()
            if str(path) in resolved_keep:
                conflicts.append(candidate["path"])
            if any(p in path.parents for p in resolved_directories):
                conflicts.append(candidate["path"])
        check("Candidates do not overlap keep entries", sorted(set(conflicts)), "Checks resolved paths and ancestor directory selections; this does not approve deletion.")
        check("Candidate paths still exist", [r["path"] for r in self.candidates if not Path(r["path"]).is_file()], "Candidates are suggestions, never executable actions.")
        cache_issues = []
        for row in self.candidates:
            if row["confidence"] != "safe-generated":
                continue
            path = Path(row["path"])
            match = re.fullmatch(r"(.+)\.cpython-\d+(?:-pytest-[\d.]+)?\.pyc", path.name)
            if not match or not path.is_file() or path.stat().st_size < 16:
                cache_issues.append(str(path))
                continue
            source = path.parent.parent / (match[1] + ".py")
            magic, mode, mtime, size = struct.unpack("<4sIII", path.read_bytes()[:16])
            if not source.is_file() or magic != importlib.util.MAGIC_NUMBER or mode != 0 or mtime != (int(source.stat().st_mtime) & 0xffffffff) or size != (source.stat().st_size & 0xffffffff):
                cache_issues.append(str(path))
        check("Bytecode candidates have matching source counterparts", cache_issues, "Checks CPython/pytest headers, not a mathematical proof of semantic identity.")
        coverage_issues = []
        for row in self.coverage:
            for audit in row["audit_ids"].split(";"):
                if audit not in self.audits:
                    coverage_issues.append(f"Unknown audit {audit}")
            for path in [row["source_path"], *filter(None, row["graphic_inputs"].split(";"))]:
                if not Path(path).is_file():
                    coverage_issues.append(path)
        check("Paper mappings resolve to reports and assets", coverage_issues, "Checks mapped sources and assets, not a fresh PDF compilation.")
        figures = sum(r["kind"] == "figure" for r in self.coverage)
        tables = sum(r["kind"] == "table" for r in self.coverage)
        check("Coverage contains 30 figures and 10 tables", [] if (figures, tables) == (30, 10) else [(figures, tables)], "Counts refer to the audited paper snapshot.")
        observed_labels = []
        for source in dict.fromkeys(row["source_path"] for row in self.coverage):
            text = re.sub(r"(?<!\\)%[^\n]*", "", Path(source).read_text())
            for match in re.finditer(r"\\begin\{(figure\*?|wrapfigure|table\*?)\}(.*?)\\end\{\1\}", text, re.S):
                observed_labels.append(";".join(re.findall(r"\\label\{([^}]+)\}", match[2])))
        check("Mapped labels match the current source floats", [] if Counter(observed_labels) == Counter(r["labels"] for r in self.coverage) else [{"observed": observed_labels}], "Detects added, removed or relabelled figures and tables in the mapped source files.")
        status_counts = dict(Counter(r["status"] for r in self.code))
        expected = self.previous["code_status_counts"]
        check("Headline counts agree with the saved validation", [] if status_counts == expected and len(self.needed) == self.previous["needed_files"] else [{"actual": status_counts, "saved": expected}], "Consistency check only; it is not a completeness certificate.")
        refs = {key for row in self.needed for key in row["audit_ids"].split(";")}
        check("Keep-list audit references resolve", sorted(refs - self.audits.keys()), "Each keep entry leads to a detailed report.")
        return {"checked_at": datetime.now(timezone.utc).isoformat(), "checks": checks, "passed": all(c["passed"] for c in checks), "limits": ["664 inventoried code paths remain unresolved.", "43 directory paths use subset-selection rules, not complete per-file manifests.", "Historical code versions and some dynamic dependencies remain unresolved.", "This review checks the audit records; it does not rerun all experiments or certify every scientific claim.", "New viewer and test files are outside the original September 7 code inventory."]}

    def summary(self) -> dict:
        changed = []
        for name, expected in self.fingerprints.items():
            path = self.directory / name
            if not path.exists() or (path.stat().st_size, path.stat().st_mtime_ns) != expected:
                changed.append(name)
        return {"read_only": True, "loaded_at": self.loaded_at, "input_digest": self.digest, "changed_inputs_since_load": changed, "audit_directory": str(self.directory), "agent_audits": sum(bool(re.fullmatch(r"e\d{2}", k)) for k in self.audits), "root_supplements": sum(not bool(re.fullmatch(r"e\d{2}", k)) for k in self.audits), "needed_files": len(self.needed), "code_files": len(self.code), "status_counts": dict(Counter(r["status"] for r in self.code)), "directory_paths": len({r["path"] for r in self.selections}), "source_candidates": sum(r["confidence"] == "candidate-only" for r in self.candidates), "cache_candidates": sum(r["confidence"] == "safe-generated" for r in self.candidates), "review_passed": self.review["passed"], "figures": sum(r["kind"] == "figure" for r in self.coverage), "tables": sum(r["kind"] == "table" for r in self.coverage)}

    def rows(self, dataset: str, query: str = "", status: str = "", audit: str = "", offset: int = 0, limit: int = 100) -> dict:
        if dataset not in {"code", "needed"} or offset < 0 or not 1 <= limit <= 250:
            raise ValueError("Invalid dataset or pagination")
        source = self.code if dataset == "code" else self.needed
        terms = query.casefold().split()
        rows = [r for r in source if all(t in r["path"].casefold() for t in terms) and (not status or r.get("status") == status) and (not audit or audit in r["audit_ids"].split(";"))]
        return {"total": len(rows), "offset": offset, "limit": limit, "rows": rows[offset:offset + limit]}

    def detail(self, path: str) -> dict:
        if path not in self.needed_by_path and path not in self.code_by_path and path not in self.candidate_by_path:
            raise KeyError("Path is not in the audit inventory")
        return {"path": path, "code": self.code_by_path.get(path), "keep": self.needed_by_path.get(path), "candidate": self.candidate_by_path.get(path), "evidence": self.evidence.get(path, []), "exists_now": Path(path).exists()}


def make_handler(catalog: AuditCatalog):
    class Handler(BaseHTTPRequestHandler):
        def respond(self, body: bytes, mime: str, status: int = 200, filename: str | None = None):
            self.send_response(status)
            self.send_header("Content-Type", mime)
            self.send_header("Content-Length", str(len(body)))
            self.send_header("Cache-Control", "no-store")
            self.send_header("X-Content-Type-Options", "nosniff")
            self.send_header("Content-Security-Policy", "default-src 'self'; script-src 'self'; style-src 'self' 'unsafe-inline'; frame-ancestors 'none'; base-uri 'none'")
            if filename:
                self.send_header("Content-Disposition", f'attachment; filename="{filename}"')
            self.end_headers()
            self.wfile.write(body)

        def do_GET(self):
            if self.headers.get("Host", "").split(":")[0] not in {"127.0.0.1", "localhost"}:
                self.respond(b"Loopback Host required", "text/plain", 403)
                return
            parsed = urlsplit(self.path)
            params = parse_qs(parsed.query)
            arg = lambda key, default="": params.get(key, [default])[0]
            try:
                if parsed.path in {"/", "/app.js"}:
                    name, mime = ("paper_code_audit_viewer.html", "text/html; charset=utf-8") if parsed.path == "/" else ("paper_code_audit_viewer.js", "text/javascript; charset=utf-8")
                    self.respond((ASSETS / name).read_bytes(), mime)
                    return
                if parsed.path == "/download":
                    name = arg("name")
                    if name not in catalog.documents:
                        raise KeyError("Unknown report")
                    self.respond(catalog.documents[name], "application/octet-stream", filename=name)
                    return
                if parsed.path == "/api/status":
                    data = catalog.summary()
                elif parsed.path == "/api/review":
                    data = catalog.review
                elif parsed.path == "/api/audits":
                    data = [{"id": k, "result": d["result"], "summary": d["summary"], "needed_entries": len(d["needed"]), "open_questions": len(d["unresolved"])} for k, d in catalog.audits.items()]
                elif parsed.path == "/api/audit":
                    data = catalog.audits[arg("id")]
                elif parsed.path == "/api/files":
                    data = catalog.rows(arg("dataset", "code"), arg("q"), arg("status"), arg("audit"), int(arg("offset", "0")), int(arg("limit", "100")))
                elif parsed.path == "/api/file":
                    data = catalog.detail(arg("path"))
                elif parsed.path == "/api/candidates":
                    data = catalog.candidates
                elif parsed.path == "/api/coverage":
                    data = catalog.coverage
                elif parsed.path == "/api/selections":
                    data = catalog.selections
                elif parsed.path == "/api/report":
                    name = arg("name", "report.md")
                    if name not in catalog.documents or not name.endswith(".md"):
                        raise KeyError("Unknown report")
                    data = {"name": name, "text": catalog.documents[name].decode()}
                else:
                    raise KeyError("Unknown endpoint")
                self.respond(json.dumps(data).encode(), "application/json; charset=utf-8")
            except (KeyError, ValueError) as error:
                self.respond(json.dumps({"error": str(error)}).encode(), "application/json", 400 if isinstance(error, ValueError) else 404)

    return Handler


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--audit-dir", type=Path, default=DEFAULT_AUDIT)
    parser.add_argument("--host", choices=["127.0.0.1"], default="127.0.0.1")
    parser.add_argument("--port", type=int, default=8003)
    args = parser.parse_args()
    catalog = AuditCatalog(args.audit_dir)
    server = ThreadingHTTPServer((args.host, args.port), make_handler(catalog))
    print(json.dumps({"url": f"http://{args.host}:{args.port}", **catalog.summary()}), flush=True)
    server.serve_forever()


if __name__ == "__main__":
    main()
